File: web/systems/issues/services/Issues.js
(function() {
		'use strict'
		angular
			.module('issues')
			.service('Issues', Issues);

		Issues.inject = ['Login', '$window', '$q'];

		function Issues(Login, $window, $q) {

		  this.getMyIssues = getMyIssues;
		  this.getOfficesIssues = getOfficesIssues;
		  this.getAssignedIssues = getAssignedIssues;
		  this.findById = findById;
			this.findByIds = findByIds;
			this.findAll = findAll;
		  this.create = create;
		  this.createComment = createComment;
		  this.changeStatus = changeStatus;
			this.changePriority = changePriority;
			this.getOffices = getOffices;
			this.getAreas = getAreas;
			this.getOfficeSubjects = getOfficeSubjects;
			this.subscribe = subscribe;
			this.searchUsers = searchUsers;
			this.updateIssue = updateIssue;

			function subscribe(event, func) {
				Login.getPrivateTransport().subscribe(event, func);
			}

		  function searchUsers(regex) {
				return Login.getPrivateTransport().call('issues.search_users', [regex]);
			}


		  function getOfficesIssues() {
				return Login.getPrivateTransport().call('issues.get_offices_issues');
			}

			/*
				busca todos los ids de los pedidos de acuerdo a los filtros pasados como parámetro.
				statuses = lista de ids de estados
				froms = lista de ids de oficinas de origen de los pedidos
				tos = lista de ids de oficinas de destino de los pedidos
			*/
			function findAll(statuses, froms, tos) {
				return Login.getPrivateTransport().call('issues.find_all', [statuses, froms, tos]);
			}

		  function getAssignedIssues(statuses, froms, tos) {
				return Login.getPrivateTransport().call('issues.get_assigned_issues', [statuses, froms, tos]).then(_findFromCacheOrLoad).then(_cacheIssues);
			}

			function getMyIssues(statuses, froms, tos) {
				return Login.getPrivateTransport().call('issues.get_my_issues', [statuses, froms, tos]).then(_findFromCacheOrLoad).then(_cacheIssues);
			}

			function updateIssue(id, status, priority) {
				var item = JSON.parse($window.sessionStorage.getItem(id));
				if (item != null) {
					item.statusId = status;
					item.priority = priority;
					$window.sessionStorage.setItem(id, JSON.stringify(item));
				}

			}

		  function findById(id) {
				if (id == null) {
					return $q.when(null);
				}
				return Login.getPrivateTransport().call('issues.find_by_id', [id]);
			}

			function findByIds(ids) {
				if (ids.length <= 0) {
					return $q.when([]);
				}
				return Login.getPrivateTransport().call('issues.find_by_ids', [ids]);
			}



		  function create(subject, description, parentId, officeId, fromOfficeId, authorId, files) {
				return Login.getPrivateTransport().call('issues.create', [subject, description, parentId, officeId, fromOfficeId, authorId, files]).then(_asyncInvalidateCache)
			}

		  function createComment(subject, description, parentId, officeId, files) {
				return Login.getPrivateTransport().call('issues.create_comment', [subject, description, parentId, officeId, files]).then(_asyncInvalidateCache);
			}

		  function changeStatus(issue, status) {
				return Login.getPrivateTransport().call('issues.change_status', [issue, status]).then(_asyncInvalidateSingleIssueCache);
			}

		  function changePriority(issue, priority) {
				return Login.getPrivateTransport().call('issues.change_priority', [issue, priority]).then(_asyncInvalidateSingleIssueCache);
			}

			function getOffices() {
				return Login.getPrivateTransport().call('issues.get_offices',[]);
			}

			function getOfficeSubjects(officeId) {
				return Login.getPrivateTransport().call('issues.get_office_subjects',[officeId]);
			}

			function getAreas(officeId) {
				return Login.getPrivateTransport().call('issues.get_areas',[officeId]);
			}


			//////////////////////////// CAPA CACHE DEL SISTEMA EN EL CLIENTE ///////////////////////
			/// la cache se compone de un indice que cachea el findAll y un cache de cada entrada del indice.
			///
			/// assignedIssues --> cachea la lista de ids de issues asignados a la persona logueada que existen en el sistema
			///
			///  issue.id ---> cache de cada issue en particular dentro del cliente
			///
			///////

			/*
				Obtiene el issue desde la cache o lo carga usando findByIds
				param: [ id1, id2, id3, .... ]
				return: [ Issue, Issue, Issue, ..... ]
			*/
			function _findFromCacheOrLoad(ids) {
				var remainingIds = [];

				// cargo desde la cache los issues existentes
				var issues = [];
				for (var i = 0; i < ids.length; i++) {
					var iss = $window.sessionStorage.getItem(ids[i]);
					if (iss != null) {
						issues.push(JSON.parse(iss));
					} else {
						remainingIds.push(ids[i]);
					}
				}

				// si estaban todos los ids pedidos en la cache
				if (remainingIds.length <= 0) {
					return $q.when(issues);
				}

				// si no, cargo lso que faltan desde el servidor usando findByIds, los cacheo y retorno tda la lista.
				var d = $q.defer();
				findByIds(remainingIds).then(_cacheIssues).then(function(iss) {
					var totalIssues = issues.concat(iss);
					return d.resolve(totalIssues);
				});
				return d.promise;
			}



			/*
				Cachea la lista de ids de issues que estan asignados a la persona
			*/
			function _asyncCacheAssignedIssuesIndex(issueIds) {
				$window.sessionStorage.setItem('assignedIssues', JSON.stringify(issueIds));
				$q.when(issueIds)
			}

			/*
				Cachea la lista de ids de issues que realizó la persona
			*/
			function _asyncCacheMyIssuesIndex(issueIds) {
				$window.sessionStorage.setItem('myIssues', JSON.stringify(issueIds));
				$q.when(issueIds)
			}


			/*
				Cachea los issues dentro de la cache, solo si ya no existe.
				param: [ Issue, Issue, ... ]
				return: [ Issue, Issue, Issue, ... ]
			*/
			function _cacheIssues(issues) {
				for (var i = 0; i < issues.length; i++) {
					var issue = issues[i];
					if ($window.sessionStorage.getItem(issue.id) == null) {
						$window.sessionStorage.setItem(issue.id, JSON.stringify(issue));
					}
				}
				return $q.when(issues);
			}

			/*
				Invalida la cache de los issues asignados a la persona y los que realizó.
			*/
			function _asyncInvalidateCache(issueId) {
				try {
					if ($window.sessionStorage.getItem('assignedIssues') != null) {
						$window.sessionStorage.removeItem('assignedIssues');
					}
					if ($window.sessionStorage.getItem('myIssues') != null) {
						$window.sessionStorage.removeItem('myIssues');
					}
					;
				} catch(err) {
					console.log(err);
				}
				return $q.when(issueId);
			}

			/*
				Invalida un issue en particular.
			*/
			function _asyncInvalidateSingleIssueCache(issue) {
				try {
					if ($window.sessionStorage.getItem(issue.id) != null) {
						$window.sessionStorage.removeItem(issue.id);
					}
				} catch(err) {
					console.log(err);
				}
				return $q.when(issue);
			}

			//////////////////////////////////


		}

})();

File: server/actions/systems/issues/issues.py
# -*- coding: utf-8 -*-
import inject
import json
import re
import logging
import psycopg2

from twisted.internet.defer import inlineCallbacks

from model.issue.issue import Issue, RedmineAPI, Attachment, IssueModel
from model.offices.office import Office
from model.users.users import User

#import asyncio
#from asyncio import coroutine
#from autobahn.asyncio.wamp import ApplicationSession

#from model.serializer.utils import  JSONSerializable

import autobahn
import wamp


class Issues(wamp.SystemComponentSession):

    conn = wamp.getConnectionManager()

    def getRegisterOptions(self):
        return autobahn.wamp.RegisterOptions(details_arg='details')

    @autobahn.wamp.register('issues.get_my_issues')
    def getMyIssues(self, statuses, froms, tos, details):
        """
            Obtiene los issues que realizó la oficina de la persona.
            TODO: implementar los filtros.
        """
        con = self.conn.get()
        try:
            userId = self.getUserId(con, details)
            return Issue.getMyIssues(con, userId)
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.get_offices_issues')
    def getOfficesIssues(self, details):
        con = self.conn.get()
        try:
            userId = self.getUserId(con, details)
            oIds = Office.getOfficesByUser(con, userId, False)
            return Issue.getOfficesIssues(con, oIds)
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.get_assigned_issues')
    def getAssignedIssues(self, statuses, froms, tos, details):
        """
            Retorna los issues asignados a las oficinas a las que pertenece la persona.
        """
        con = self.conn.get()
        try:
            logging.info(statuses)
            logging.info(froms)
            logging.info(tos)

            if statuses is None:
                return []
            assert isinstance(statuses,list)
            if len(statuses) <= 0:
                return []

            userId = self.getUserId(con, details)
            oIds = Office.findByUser(con, userId, types=None, tree=True)
            toIds = [oid for oid in oIds if oid in tos]
            return Issue.getAssignedIssues(con, userId, toIds, froms, statuses)
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.find_by_id')
    def findById(self, issueid, details):
        con = self.conn.get()
        try:
            issues = Issue.findByIds(con, [issueid])
            if issues is None or len(issues) <= 0:
                return None
            return issues[0]
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.find_by_ids')
    def findByIds(self, issuesIds, details):
        con = self.conn.get()
        try:
            logging.info(issuesIds)
            if len(issuesIds) <= 0:
                return []
            return Issue.findByIds(con, issuesIds)
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.create')
    @inlineCallbacks
    def create(self, subject, description, parentId, officeId, fromOfficeId, authorId, files, details):
        con = self.conn.get()
        try:
            print('create issue')
            userId = self.getUserId(con, details)
            authorId = userId if authorId is  None else authorId
            tracker = IssueModel.TRACKER_ERROR
            issueId = IssueModel.create(con, parentId, officeId, authorId, subject, description, fromOfficeId, userId, files, tracker)
            con.commit()
            print(issueId)
            yield self.publish('issues.issue_created_event', issueId, authorId, fromOfficeId, officeId)
            return issueId
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.create_comment')
    @inlineCallbacks
    def createComment(self, subject, description, parentId, projectId, files, details):
        con = self.conn.get()
        try:
            self.log.info('createComment')
            userId = self.getUserId(con, details)
            tracker = IssueModel.TRACKER_COMMENT
            issueId = IssueModel.create(con, parentId, projectId, userId, subject, description, '', '', files, tracker)
            con.commit()
            yield self.publish('issues.comment_created_event', parentId, issueId)
            return issueId

        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.change_status')
    @inlineCallbacks
    def changeStatus(self, issue, status, details):
        iss = issue.changeStatus(status)
        yield self.publish('issues.updated_event', issue.id, status, issue.priority)
        return issue

    @autobahn.wamp.register('issues.change_priority')
    @inlineCallbacks
    def changePriority(self, issue, priority, details):
        iss = issue.changePriority(priority)
        yield self.publish('issues.updated_event', issue.id, issue.statusId, priority)
        return issue

    @autobahn.wamp.register('issues.get_offices')
    def getOffices(self, details):
        con = self.conn.get()
        try:
            userId = self.getUserId(con, details)
            return IssueModel.getOffices(con, userId)
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.search_users')
    def searchUsers(self, regex, details):
        con = self.conn.get()
        try:
            return IssueModel.searchUsers(con, regex)
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.get_office_subjects')
    def getOfficeSubjects(self, officeId, details):
        con = self.conn.get()
        try:
            return IssueModel.getSubjectTypes(con, officeId)
        finally:
            self.conn.put(con)

    @autobahn.wamp.register('issues.get_areas')
    def getAreas(self, oId, details):
        con = self.conn.get()
        try:
            return IssueModel.getAreas(con, oId)
        finally:
            self.conn.put(con)
